### ISSUE TITLE
[LibOS] Emulation of sendfile() has bad performance

### ISSUE THREAD

**Issue by dimakuv** [oscarlab/graphene#2025]
## Description of the problem

This is the Nginx performance problem. There is sub-optimal implementation of `sendfile()` syscall in Graphene: https://github.com/oscarlab/graphene/blob/master/LibOS/shim/src/sys/shim_fs.c#L610

Particular lines that are bad for performance:
-	https://github.com/oscarlab/graphene/blob/master/LibOS/shim/src/sys/shim_fs.c#L634
-	https://github.com/oscarlab/graphene/blob/master/LibOS/shim/src/sys/shim_fs.c#L644

This is what native Nginx does for each HTTP request:
```
[pid 25977] openat(AT_FDCWD, "/home/dimakuv/graphene/Examples/nginx/install/html/random/10K.1.html"…) = 12
[pid 25977] fstat(12, {st_mode=S_IFREG|0664, st_size=10240, ...}) = 0
[pid 25977] writev(11, [{iov_base="HTTP/1.1 200 OK\r\nServer: nginx/1"..., iov_len=241}], 1) = 241
[pid 25977] sendfile(11, 12, [0] => [10240], 10240) = 10240   // !!!!!
[pid 25977] write(4, "127.0.0.1 - - [11/Dec/2020:04:40"..., 109) = 109
[pid 25977] close(12)                   = 0
```

In other words, Nginx opens a file-to-send-to-client and sends the whole file contents via `sendfile()` on the client socket. `sendfile()` is a simple atomic operation in Linux kernel, but Graphene has a poor re-implementation of it. Graphene sees that offset must be updated and does it via handle-specific `seek()`. But `seek()` in turn performs host-level `ppoll()` and `fcntl()` and `ioctl()`…

The correct solution is to track and cache the modified offset inside the Graphene’s LibOS layer instead of actively quering/setting it every time. I will try to come up with a fix for it next week.

## Steps to reproduce

Just run our Nginx example.

**Comment by dimakuv** [comment]
Ok, it's more complicated than that.

Nginx web server always does `open("html-file-to-send") + sendfile() + close()`, for each request. Since the HTML file is always opened and closed, Nginx doesn't cache the file's contents, even if the file is completely static and never changes (as in our benchmark case). In fact, this is a right thing to do: the Linux kernel will cache the file contents anyway. See also https://stackoverflow.com/questions/38813399/cache-a-static-file-in-memory-forever-on-nginx.

But Graphene-SGX doesn't have any other option but to open the file, mmap and copy it inside the enclave, sendfile, and then munmap and close it. (Graphene-SGX cannot rely on the page cache of the host Linux kernel.) **This is where the overhead comes from.**

One *hacky* way to compare native vs Graphene performance more apples-to-apples is to **disable sendfile** in the config file: http://nginx.org/en/docs/http/ngx_http_core_module.html#sendfile. In this case, Nginx starts doing explicit `read()` from the HTML file instead of relying on the Linux kernel:
```
[pid 26192] openat(AT_FDCWD, "/home/dimakuv/graphene/Examples/nginx/install/html/random/10K.1.html", O_RDONLY|O_NONBLOCK) = 12
[pid 26192] fstat(12, {st_mode=S_IFREG|0664, st_size=10240, ...}) = 0
[pid 26192] pread64(12, "\200\264\363\317\322\254\244\255\257\246?\275S,5U\342\24\355\16\237u\335Kb*=M\206\201\300 "..., 10240, 0) = 10240
[pid 26192] writev(11, [{iov_base="HTTP/1.1 200 OK\r\nServer: nginx/1"..., iov_len=241}, {iov_base="\200\264\363\317\322\254\244\255\257\246?\275S,5U\342\24\355\16\237u\335Kb*=M\206\201\300 "..., iov_len=10240}], 2) = 10481
[pid 26192] write(4, "127.0.0.1 - - [11/Dec/2020:05:25"..., 109) = 109
[pid 26192] close(12)                   = 0
```

**Comment by dimakuv** [comment]
Some very quick results on Graphene-nonsgx vs native, with `sendfile = on` and `sendfile = off`:
```
------ native, sendfile=on [default]
Requests per second:    154135.45 [#/sec] (mean)
Time per request:       0.208 [ms] (mean)
Transfer rate:          1577623.14 [Kbytes/sec] received

------ graphene, sendfile=on [default]
Requests per second:    129157.92 [#/sec] (mean)
Time per request:       0.248 [ms] (mean)
Transfer rate:          1321087.56 [Kbytes/sec] received     // 19.4% overhead

------ native, sendfile=off
Requests per second:    172597.70 [#/sec] (mean)
Time per request:       0.185 [ms] (mean)
Transfer rate:          1766589.83 [Kbytes/sec] received

------ graphene, sendfile=off
Requests per second:    163826.73 [#/sec] (mean)
Time per request:       0.195 [ms] (mean)
Transfer rate:          1675696.38 [Kbytes/sec] received    // 5.4% overhead
```

By setting config option `sendfile = off`, we see that Graphene overhead drops from 20% to 5%, as expected.

**Comment by dimakuv** [comment]
The optimization of `sendfile()` syscall emulation in Graphene is tabled for now.

In an offline discussion we decided that the above scenario of HTTP client requests is not representative of real workloads, since all secure communication is done via HTTPS anyway. And when Nginx uses HTTPS, it doesn't rely on `sendfile()`.

So there are currently no real-world uses of `sendfile()` that are applicable to Graphene + SGX.

**Comment by mkow** [comment]
> Some very quick results on Graphene-nonsgx vs native, with `sendfile = on` and `sendfile = off`:

Huh, why disabling sendfile gives such a huge boost natively? Why does anyone use `sendfile = on` then?

**Comment by dimakuv** [comment]
> > Some very quick results on Graphene-nonsgx vs native, with `sendfile = on` and `sendfile = off`:
> Huh, why disabling sendfile gives such a huge boost natively? Why does anyone use `sendfile = on` then?

I think this depends on various factors. It was fast on my development machine, but when I tried on another (much more powerful machine with a different hard drive and newer Linux) the result was reverse...

**Comment by dimakuv** [comment]
This is very old. We rewrote `sendfile()` implementation recently. Anyway, `sendfile` performance is not an issue for real-world apps since they use TLS and encryption anyway.
